[2.9 ACRA] NullPointerException: File Import

First crash for 2.9.6beta4

https://couchdb.ankidroid.org/acralyzer/_design/acralyzer/index.html#/report-details/fc81b2e6-c6bb-4807-8abc-b7adfb526c20

```
java.lang.RuntimeException: Unable to start activity ComponentInfo{com.ichi2.anki/com.ichi2.anki.IntentHandler}: java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String android.net.Uri.getScheme()' on a null object reference
at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3516)
at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3668)
at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:83)
at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2193)
at android.os.Handler.dispatchMessage(Handler.java:107)
at android.os.Looper.loop(Looper.java:237)
at android.app.ActivityThread.main(ActivityThread.java:7971)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1076)
Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String android.net.Uri.getScheme()' on a null object reference
at com.ichi2.utils.ImportUtils.handleFileImport(ImportUtils.java:2)
at com.ichi2.anki.IntentHandler.onCreate(IntentHandler.java:10)
at android.app.Activity.performCreate(Activity.java:7957)
at android.app.Activity.performCreate(Activity.java:7946)
at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1307)
at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3491)
... 11 more
```

--- a/AnkiDroid/src/main/java/com/ichi2/utils/ImportUtils.java
+++ b/AnkiDroid/src/main/java/com/ichi2/utils/ImportUtils.java
@@ -5,6 +5,7 @@ import android.content.Context;
 import android.content.Intent;
 import android.database.Cursor;
 import android.net.Uri;
+import android.os.Build;
 import android.os.Bundle;
 import android.os.Message;
 import android.provider.OpenableColumns;
@@ -47,12 +48,24 @@ public class ImportUtils {
         String errorMessage = null;
 
         if (intent.getData() == null) {
-            return context.getString(R.string.import_error_unhandled_request);
+            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.JELLY_BEAN || intent.getClipData() == null) {
+                return context.getString(R.string.import_error_unhandled_request);
+            }
+            try {
+                Uri clipUri = intent.getClipData().getItemAt(0).getUri();
+                return handleContentProviderFile(context, intent, clipUri);
+            } catch (Exception e) {
+                return context.getString(R.string.import_error_exception, e.getLocalizedMessage());
+            }
         }
 
         // If the file is being sent from a content provider we need to read the content before we can open the file
         if ("content".equals(intent.getData().getScheme())) {
-            return handleContentProviderFile(context, intent, intent.getData());
+            try {
+                return handleContentProviderFile(context, intent, intent.getData());
+            } catch (Exception e) {
+                return context.getString(R.string.import_error_exception, e.getLocalizedMessage());
+            }
         } else if ("file".equals(intent.getData().getScheme())) {
             // When the VIEW intent is sent as a file, we can open it directly without copying from content provider
             String filename = intent.getData().getPath();

--- a/AnkiDroid/src/main/java/com/ichi2/utils/ImportUtils.java
+++ b/AnkiDroid/src/main/java/com/ichi2/utils/ImportUtils.java
@@ -8,6 +8,7 @@ import android.net.Uri;
 import android.os.Bundle;
 import android.os.Message;
 import android.provider.OpenableColumns;
+import android.text.TextUtils;
 
 import com.afollestad.materialdialogs.MaterialDialog;
 import com.ichi2.anki.AnkiActivity;
@@ -40,6 +41,9 @@ public class ImportUtils {
         // This intent is used for opening apkg package files
         // We want to go immediately to DeckPicker, clearing any history in the process
         Timber.i("IntentHandler/ User requested to view a file");
+        String extras = intent.getExtras() == null ? "none" : TextUtils.join(", ", intent.getExtras().keySet());
+        Timber.i("Intent: %s. Data: %s", intent, extras);
+
         String errorMessage = null;
         // If the file is being sent from a content provider we need to read the content before we can open the file
         if ("content".equals(intent.getData().getScheme())) {

--- a/AnkiDroid/src/main/java/com/ichi2/utils/ImportUtils.java
+++ b/AnkiDroid/src/main/java/com/ichi2/utils/ImportUtils.java
@@ -45,6 +45,15 @@ public class ImportUtils {
         String extras = intent.getExtras() == null ? "none" : TextUtils.join(", ", intent.getExtras().keySet());
         Timber.i("Intent: %s. Data: %s", intent, extras);
 
+        try {
+            return handleFileImportInternal(context, intent);
+        } catch (Exception e) {
+            return context.getString(R.string.import_error_exception, e.getLocalizedMessage());
+        }
+    }
+
+    //Added to remove exception handlers
+    private static String handleFileImportInternal(Context context, Intent intent) {
         String errorMessage = null;
 
         if (intent.getData() == null) {
@@ -52,22 +61,14 @@ public class ImportUtils {
             if (Build.VERSION.SDK_INT < Build.VERSION_CODES.JELLY_BEAN || intent.getClipData() == null) {
                 return context.getString(R.string.import_error_unhandled_request);
             }
-            try {
-                Uri clipUri = intent.getClipData().getItemAt(0).getUri();
-                return handleContentProviderFile(context, intent, clipUri);
-            } catch (Exception e) {
-                return context.getString(R.string.import_error_exception, e.getLocalizedMessage());
-            }
+            Uri clipUri = intent.getClipData().getItemAt(0).getUri();
+            return handleContentProviderFile(context, intent, clipUri);
         }
 
         // If the file is being sent from a content provider we need to read the content before we can open the file
         if ("content".equals(intent.getData().getScheme())) {
             Timber.i("Attempting to read content from intent.");
-            try {
-                return handleContentProviderFile(context, intent, intent.getData());
-            } catch (Exception e) {
-                return context.getString(R.string.import_error_exception, e.getLocalizedMessage());
-            }
+            return handleContentProviderFile(context, intent, intent.getData());
         } else if ("file".equals(intent.getData().getScheme())) {
             Timber.i("Attempting to read file from intent.");
             // When the VIEW intent is sent as a file, we can open it directly without copying from content provider

--- a/AnkiDroid/src/main/java/com/ichi2/utils/ImportUtils.java
+++ b/AnkiDroid/src/main/java/com/ichi2/utils/ImportUtils.java
@@ -49,6 +49,7 @@ public class ImportUtils {
             return handleFileImportInternal(context, intent);
         } catch (Exception e) {
             AnkiDroidApp.sendExceptionReport(e, "handleFileImport");
+            Timber.e(e, "failed to handle import intent");
             return context.getString(R.string.import_error_exception, e.getLocalizedMessage());
         }
     }

--- a/AnkiDroid/src/main/java/com/ichi2/utils/ImportUtils.java
+++ b/AnkiDroid/src/main/java/com/ichi2/utils/ImportUtils.java
@@ -45,6 +45,11 @@ public class ImportUtils {
         Timber.i("Intent: %s. Data: %s", intent, extras);
 
         String errorMessage = null;
+
+        if (intent.getData() == null) {
+            return context.getString(R.string.import_error_unhandled_request);
+        }
+
         // If the file is being sent from a content provider we need to read the content before we can open the file
         if ("content".equals(intent.getData().getScheme())) {
             return handleContentProviderFile(context, intent, intent.getData());

--- a/AnkiDroid/src/main/java/com/ichi2/utils/ImportUtils.java
+++ b/AnkiDroid/src/main/java/com/ichi2/utils/ImportUtils.java
@@ -47,7 +47,7 @@ public class ImportUtils {
         String errorMessage = null;
         // If the file is being sent from a content provider we need to read the content before we can open the file
         if ("content".equals(intent.getData().getScheme())) {
-            return handleContentProviderFile(context, intent);
+            return handleContentProviderFile(context, intent, intent.getData());
         } else if ("file".equals(intent.getData().getScheme())) {
             // When the VIEW intent is sent as a file, we can open it directly without copying from content provider
             String filename = intent.getData().getPath();
@@ -63,13 +63,13 @@ public class ImportUtils {
     }
 
 
-    private static String handleContentProviderFile(Context context, Intent intent) {
+    private static String handleContentProviderFile(Context context, Intent intent, Uri data) {
         // Get the original filename from the content provider URI
         String errorMessage = null;
         String filename = null;
         Cursor cursor = null;
         try {
-            cursor = context.getContentResolver().query(intent.getData(), new String[]{OpenableColumns.DISPLAY_NAME}, null, null, null);
+            cursor = context.getContentResolver().query(data, new String[]{OpenableColumns.DISPLAY_NAME}, null, null, null);
             if (cursor != null && cursor.moveToFirst()) {
                 filename = cursor.getString(0);
                 Timber.d("handleFileImport() Importing from content provider: %s", filename);

--- a/AnkiDroid/src/main/java/com/ichi2/utils/ImportUtils.java
+++ b/AnkiDroid/src/main/java/com/ichi2/utils/ImportUtils.java
@@ -43,47 +43,7 @@ public class ImportUtils {
         String errorMessage = null;
         // If the file is being sent from a content provider we need to read the content before we can open the file
         if ("content".equals(intent.getData().getScheme())) {
-            // Get the original filename from the content provider URI
-            String filename = null;
-            Cursor cursor = null;
-            try {
-                cursor = context.getContentResolver().query(intent.getData(), new String[]{OpenableColumns.DISPLAY_NAME}, null, null, null);
-                if (cursor != null && cursor.moveToFirst()) {
-                    filename = cursor.getString(0);
-                    Timber.d("handleFileImport() Importing from content provider: %s", filename);
-                }
-            } finally {
-                if (cursor != null)
-                    cursor.close();
-            }
-
-            // Hack to fix bug where ContentResolver not returning filename correctly
-            if (filename == null) {
-                if (intent.getType() != null && ("application/apkg".equals(intent.getType()) || ImportUtils.hasValidZipFile(context, intent))) {
-                    // Set a dummy filename if MIME type provided or is a valid zip file
-                    filename = "unknown_filename.apkg";
-                    Timber.w("Could not retrieve filename from ContentProvider, but was valid zip file so we try to continue");
-                } else {
-                    Timber.e("Could not retrieve filename from ContentProvider or read content as ZipFile");
-                    AnkiDroidApp.sendExceptionReport(new RuntimeException("Could not import apkg from ContentProvider"), "IntentHandler.java", "apkg import failed");
-                    errorMessage = AnkiDroidApp.getAppResources().getString(R.string.import_error_content_provider, AnkiDroidApp.getManualUrl() + "#importing");
-                }
-            }
-
-            if (!isValidPackageName(filename)) {
-                // Don't import if file doesn't have an Anki package extension
-                errorMessage = context.getResources().getString(R.string.import_error_not_apkg_extension, filename);
-            } else if (filename != null) {
-                // Copy to temporary file
-                String tempOutDir = Uri.fromFile(new File(context.getCacheDir(), filename)).getEncodedPath();
-                errorMessage = ImportUtils.copyFileToCache(context, intent, tempOutDir) ? null : "copyFileToCache() failed (possibly out of storage space)";
-                // Show import dialog
-                if (errorMessage == null) {
-                    ImportUtils.sendShowImportFileDialogMsg(tempOutDir);
-                } else {
-                    AnkiDroidApp.sendExceptionReport(new RuntimeException("Error importing apkg file"), "IntentHandler.java", "apkg import failed");
-                }
-            }
+            return handleContentProviderFile(context, intent);
         } else if ("file".equals(intent.getData().getScheme())) {
             // When the VIEW intent is sent as a file, we can open it directly without copying from content provider
             String filename = intent.getData().getPath();
@@ -93,6 +53,53 @@ public class ImportUtils {
                 ImportUtils.sendShowImportFileDialogMsg(filename);
             } else {
                 errorMessage = context.getResources().getString(R.string.import_error_not_apkg_extension, filename);
+            }
+        }
+        return errorMessage;
+    }
+
+
+    private static String handleContentProviderFile(Context context, Intent intent) {
+        // Get the original filename from the content provider URI
+        String errorMessage = null;
+        String filename = null;
+        Cursor cursor = null;
+        try {
+            cursor = context.getContentResolver().query(intent.getData(), new String[]{OpenableColumns.DISPLAY_NAME}, null, null, null);
+            if (cursor != null && cursor.moveToFirst()) {
+                filename = cursor.getString(0);
+                Timber.d("handleFileImport() Importing from content provider: %s", filename);
+            }
+        } finally {
+            if (cursor != null)
+                cursor.close();
+        }
+
+        // Hack to fix bug where ContentResolver not returning filename correctly
+        if (filename == null) {
+            if (intent.getType() != null && ("application/apkg".equals(intent.getType()) || ImportUtils.hasValidZipFile(context, intent))) {
+                // Set a dummy filename if MIME type provided or is a valid zip file
+                filename = "unknown_filename.apkg";
+                Timber.w("Could not retrieve filename from ContentProvider, but was valid zip file so we try to continue");
+            } else {
+                Timber.e("Could not retrieve filename from ContentProvider or read content as ZipFile");
+                AnkiDroidApp.sendExceptionReport(new RuntimeException("Could not import apkg from ContentProvider"), "IntentHandler.java", "apkg import failed");
+                errorMessage = AnkiDroidApp.getAppResources().getString(R.string.import_error_content_provider, AnkiDroidApp.getManualUrl() + "#importing");
+            }
+        }
+
+        if (!isValidPackageName(filename)) {
+            // Don't import if file doesn't have an Anki package extension
+            errorMessage = context.getResources().getString(R.string.import_error_not_apkg_extension, filename);
+        } else if (filename != null) {
+            // Copy to temporary file
+            String tempOutDir = Uri.fromFile(new File(context.getCacheDir(), filename)).getEncodedPath();
+            errorMessage = ImportUtils.copyFileToCache(context, intent, tempOutDir) ? null : "copyFileToCache() failed (possibly out of storage space)";
+            // Show import dialog
+            if (errorMessage == null) {
+                ImportUtils.sendShowImportFileDialogMsg(tempOutDir);
+            } else {
+                AnkiDroidApp.sendExceptionReport(new RuntimeException("Error importing apkg file"), "IntentHandler.java", "apkg import failed");
             }
         }
         return errorMessage;

--- a/AnkiDroid/src/main/java/com/ichi2/utils/ImportUtils.java
+++ b/AnkiDroid/src/main/java/com/ichi2/utils/ImportUtils.java
@@ -59,7 +59,9 @@ public class ImportUtils {
 
         if (intent.getData() == null) {
             Timber.i("No intent data. Attempting to read clip data.");
-            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.JELLY_BEAN || intent.getClipData() == null) {
+            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.JELLY_BEAN
+                    || intent.getClipData() == null
+                    || intent.getClipData().getItemCount() == 0) {
                 return context.getString(R.string.import_error_unhandled_request);
             }
             Uri clipUri = intent.getClipData().getItemAt(0).getUri();

--- a/AnkiDroid/src/main/java/com/ichi2/utils/ImportUtils.java
+++ b/AnkiDroid/src/main/java/com/ichi2/utils/ImportUtils.java
@@ -48,6 +48,7 @@ public class ImportUtils {
         try {
             return handleFileImportInternal(context, intent);
         } catch (Exception e) {
+            AnkiDroidApp.sendExceptionReport(e, "handleFileImport");
             return context.getString(R.string.import_error_exception, e.getLocalizedMessage());
         }
     }

--- a/AnkiDroid/src/main/java/com/ichi2/utils/ImportUtils.java
+++ b/AnkiDroid/src/main/java/com/ichi2/utils/ImportUtils.java
@@ -48,6 +48,7 @@ public class ImportUtils {
         String errorMessage = null;
 
         if (intent.getData() == null) {
+            Timber.i("No intent data. Attempting to read clip data.");
             if (Build.VERSION.SDK_INT < Build.VERSION_CODES.JELLY_BEAN || intent.getClipData() == null) {
                 return context.getString(R.string.import_error_unhandled_request);
             }
@@ -61,12 +62,14 @@ public class ImportUtils {
 
         // If the file is being sent from a content provider we need to read the content before we can open the file
         if ("content".equals(intent.getData().getScheme())) {
+            Timber.i("Attempting to read content from intent.");
             try {
                 return handleContentProviderFile(context, intent, intent.getData());
             } catch (Exception e) {
                 return context.getString(R.string.import_error_exception, e.getLocalizedMessage());
             }
         } else if ("file".equals(intent.getData().getScheme())) {
+            Timber.i("Attempting to read file from intent.");
             // When the VIEW intent is sent as a file, we can open it directly without copying from content provider
             String filename = intent.getData().getPath();
             Timber.d("Importing regular file: %s", filename);

--- a/AnkiDroid/src/main/res/values/02-strings.xml
+++ b/AnkiDroid/src/main/res/values/02-strings.xml
@@ -152,6 +152,7 @@
     <string name="import_message_replace_confirm">This will delete your existing collection and replace it with the data of file %s</string>
     <string name="import_message_add_confirm">Add “%s” to collection? This may take a long time</string>
     <string name="import_log_no_apkg">This isn’t a valid Anki package file</string>
+    <string name="import_error_unhandled_request">Unable to process import request</string>
     <string name="import_error_not_apkg_extension">Filename “%s” doesn’t have .apkg or .colpkg extension</string>
     <string name="import_error_content_provider">The selected file couldn’t be imported automatically by AnkiDroid. Please see the user manual for how to manually import anki files: \n%s</string>
     <string name="import_replacing">Replacing collection…</string>

--- a/AnkiDroid/src/main/res/values/02-strings.xml
+++ b/AnkiDroid/src/main/res/values/02-strings.xml
@@ -153,6 +153,7 @@
     <string name="import_message_add_confirm">Add “%s” to collection? This may take a long time</string>
     <string name="import_log_no_apkg">This isn’t a valid Anki package file</string>
     <string name="import_error_unhandled_request">Unable to process import request</string>
+    <string name="import_error_exception">Failed to import package\n\n%s</string>
     <string name="import_error_not_apkg_extension">Filename “%s” doesn’t have .apkg or .colpkg extension</string>
     <string name="import_error_content_provider">The selected file couldn’t be imported automatically by AnkiDroid. Please see the user manual for how to manually import anki files: \n%s</string>
     <string name="import_replacing">Replacing collection…</string>

